lib/device.js

```diff
@@ -1,206 +1,173 @@
-   * Returns null if message is invalid. Returns the parsed JavaScript
-   * object if the message is valid.
-   */
-  var handleMessage = function(topic, message) {
-
-    debug('Received MQTT message on topic ' + topic);
-
-    // The only topics this client supports are commands. Something else
-    // was sent, so just bail.
-    if(topic !== commandTopic) {
-      return null;
-    }
-
-    message = parseMessage(message);
-    if(!message) {
-      return null;
-    }
-
-    debug('Received MQTT message:');
-    debug(JSON.stringify(message));
-
-    device.emit('command', message);
-
-    return message;
-  };
-
-  /**
-   * Parses an incoming mqtt message from the Losant broker..
-   * Returns an object or null.
-   * message: the incoming Buffer to parse and validate.
-   */
-  var parseMessage = function(message) {
-
-    // Attempt to parse the message.
-    try { message = EJSON.parse(message.toString()); }
-    catch(e) { return null; }
-
-    // The message has to be something.
-    if(!message) {
-      return null;
-    }
-
-    return message;
-  };
-
-  /**
-   * Sends device state to Losant.
-   * deviceId: the ID of the device to report state for.
-   * state: the state to send. Typically an object with { attribute: value }
-   * time (optional): the time at which the state occurred (defaults to now)
-   * cb: the callback handling the response.
-   */
-  device.sendState = function(state, time, callback) {
-
-    if(typeof(time) === 'function') {
-      callback = time;
-      time = null;
-    }
-
-    time = time || new Date();
-    var payload = { time: time, data: state };
-
-    debug('Publishing state:');
-    debug(EJSON.stringify(payload));
-
-    // May not be connected. Skip publish if not.
-    if(mqttClient) {
-      mqttClient.publish(stateTopic, EJSON.stringify(payload), callback);
-    }
-    else {
-      if(callback) { callback(null); }
-    }
-
-    return { topic: stateTopic, payload: payload };
-  };
-
-  return device;
+  this.commandTopic = util.format(config.topicFormatCommand, options.id);
+  this.stateTopic = util.format(config.topicFormatState, options.id);
+
+  this.mqtt = options.mqtt || new MQTT();
+  this.mqtt.on('connect', this.handleConnect.bind(this));
+  this.mqtt.on('message', this.handleMessage.bind(this));
+
+  // Attach all events.
+  this.mqtt.on('connect', (function() {
+    this.emit('connect');
+  }).bind(this));
+
+  this.mqtt.on('reconnect', (function() {
+    this.emit('reconnect');
+  }).bind(this));
+
+  this.mqtt.on('close', (function() {
+    this.emit('close');
+  }).bind(this));
+
+  this.mqtt.on('offline', (function() {
+    this.emit('offline');
+  }).bind(this));
+
+  this.mqtt.on('error', (function(err) {
+    this.emit('error', err);
+  }).bind(this));
+}
+
+util.inherits(Device, EventEmitter);
+
+/**
+ * Returns whether or not the device is connected.
+ */
+Device.prototype.isConnected = function() {
+  return this.mqtt.client ? this.mqtt.client.connected : false;
```

Comment:
it seems weird to be digging into the mqtt object and touching and checking client all the time - for this, for subscribe, for publish, for disconnect - should they be wrapped up nicer in the mqtt obj?